articles/create-a-node-cli/index.md

```diff
@@ -0,0 +1,264 @@
+
+const axios = require("axios");
+const chalk = require("chalk");
+
+let url = "";
+
+function getJoke() {
+  url = "https://official-joke-api.appspot.com/random_joke";
+ axios({
+ method: "get",
+ url: url,
+ })
+ .then((res) => {
+ const setup = chalk.cyan(res.data.setup);
+ const punchline = chalk.green(res.data.punchline);
+ console.log(`${setup} - ${punchline}`);
+ })
+ .catch((err) => {
+ const log = chalk.red(err);
+ console.log(log);
+ });
+}
+
+function getQuote() {
+axios({
+    method: 'get',
+    url: "http://quotes.stormconsultancy.co.uk/random.json",
+}).then((res)=> {
+    const randomQuote = chalk.cyan(res.data.quote);
+    const randomAuthor = chalk.green(res.data.author);
+    const log = `${randomQuote} - ${randomAuthor}`;
+    console.log(log);
+}).catch((err)=>{
+    const errorLog = chalk.red(err);
+    console.log(errorLog);
+})
+```
+
+Note: We are using chalk to color our logs and get a colorful output. We setup an empty string called `url` so that when the user selects quotes, we load in the quote API URL and the same goes for jokes.
+
+For now, the code won't work because we're not calling any of the functions.
+
+## Understanding option/command/argument parsing
+
+A good example of option parsing is when we passed _-y_ to our _npm init_ command to set everything to default.
+
+The commands used in CLI applications often come with other options or flags. They extend the functionality of the application like in the example above.
+
+A common application of options is when asking for user input which determines the output of the application.
+
+Node.js comes with an in-built app interface for the process module called `process.argv`. It takes in the arguments passed, invokes a method and returns an array with all the arguments passed into the main process.
+
+Using _process.arg_ might be tedious and not beginner friendly so we will use [yargs](https://www.npmjs.com/package/yargs) to help us process the options.
+
+Yargs offers a lot of options that make it very powerful and widely used. Some of them include `.command()`, `.help()`, `.example()`, `.usage()` and others. To read more on the different capabilities of the options, see the yargs [docs](https://github.com/yargs/yargs).
+
+We will be making use of the command option to allow us to invoke a function when the set command is triggered. The syntax for this is:
+
+```javascript
+const argv = require("yargs");
+argv.command(
+  "name of command",
+  "description of the command",
+  functionA,
+  functionB
+).argv;
+```
+
+Note: _functionA_ will define the command arguments, then _functionB_ will implement the function.
+
+In our project we need two commands, let's set them up
+
+```javascript
+const argv = require("yargs");
+argv.command(
+  "joke",
+  "Fetching your joke",
+  (yargs) => {},
+  (argv) => {
+    getJoke();
+  }
+).argv;
+
+argv.command(
+  "quote",
+  "Fetching a quote",
+  (yargs) => {},
+  (argv) => {
+    getQuotes();
+  }
+).argv;
```

Comment:
@kimkimani, thanks for pushing your changes. 

Running the full code example works but combining the code snippets you've added before (https://pastebin.com/FSadkewj) seems to result in unexpected end of input for both quote and joke. Are there any differences?

Reply:
Oh yeah @louisefindlay23 , The reason is that passing the url in a string before hand results in that error.
I have fixed that tiny mishap.

Reply:
@katungi, tried using your latest code but it still results in that error.

Reply:
Hello @louisefindlay23. That has been caused by the version of node you are using. I noticed that yargs has deprecated support for version 8 and below. I have added a note to let readers know of this before hand

Reply:
I'm using v12.16.1. I used Git to track the differences between your final code snippet and the two used earlier in your tutorial (which combined) should produce the same result and they are different. 

@katungi, you should test the code snippet below "We will create two functions that will do the fetching using axios." and the one below "In our project we need two commands, let's set them up" together. They produce the EOF error.

Reply:
I have diagnosed the real problem and fixed it. Hopefully, that fixes it. I had missed a curly brace @louisefindlay23